### PR TITLE
feat: wrap document individually with --batched flag

### PR DESCRIPTION
**Pull request by Nebulis**
also adding higher timeout for some test due to ethers throttling

### CODE REVIEW

File: src/implementations/wrap/wrap.ts [hunk 1]
```diff
@@ -202,10 +213,15 @@ export const wrap = async ({
   const hashMap = merkleHashmap(individualDocumentHashes);
```
Comment (yehjxraymond) [inline src/implementations/wrap/wrap.ts:213]: Actually this is one of the most expensive operation. We can skip by this entirely. 

I'm thinking if we should just have another function to handle non-batched documents, its quite straightforward and doesnt need to copy to intermediate directory.... Can straightaway copy individual files over from input dir to output dir, running the batch function individually.

Comment (yehjxraymond) [reply src/implementations/wrap/wrap.ts:213]: In fact after phase 1, phase 2 and 3 is not needed anymore...

Comment (Nebulis) [reply src/implementations/wrap/wrap.ts]: changed =)

**Review by yehjxraymond — COMMENTED** [review]
The current method will result in 2 sets of IO. The original implementation is to optimise on memory by paging the content of the file to disk and then perform the merkle tree building in memory. 

There is no need for intermediate directory when not batching them, saving one set of IO.

File: src/implementations/wrap/wrap.ts [hunk 2]
```diff
@@ -193,19 +219,40 @@ export const wrap = async ({
...
+      outputPathType,
+      digestedDocumentPath: outputPath,
+    });
+  } else {
+    const documentFileNames = await documentsInDirectory(intermediateDir);
+    documentFileNames.forEach((file) => {
+      const document = readOpenAttestationFile(file);
+      writeOutput({ outputPathType, digestedDocumentPath: outputPath, file, document });
+    });
+  }
```
Comment (yehjxraymond) [inline src/implementations/wrap/wrap.ts:253]: does this still needs to be ran?

File: src/implementations/wrap/wrap.ts [hunk 3]
```diff
@@ -193,19 +219,40 @@ export const wrap = async ({
...
+    schema
+  );
 
   if (!individualDocumentHashes || individualDocumentHashes.length === 0)
     throw new Error(`No documents found in ${inputPath}`);
 
-  // Phase 2: Efficient merkling to build hashmap
-  const hashMap = merkleHashmap(individualDocumentHashes);
+  let merkleRoot: string | undefined;
+  if (batched) {
```
Comment (yehjxraymond) [inline src/implementations/wrap/wrap.ts:236]: Cant we return if its not batched?

File: src/implementations/wrap/wrap.ts [hunk 4]
```diff
@@ -193,19 +219,40 @@ export const wrap = async ({
 
   // Phase 1: For each document, read content, digest and write to file
   const schema = await loadSchema(schemaPath);
-  const individualDocumentHashes = await digestDocument(inputPath, intermediateDir, version, unwrap, schema);
+  const individualDocumentHashes = await digestDocument(
```
Comment (yehjxraymond) [inline src/implementations/wrap/wrap.ts:222]: Might want to rename `digestDocument` to something like `wrapIndividualDocumentsToDir`

File: src/implementations/wrap/wrap.ts [hunk 5]
```diff
@@ -193,19 +219,40 @@ export const wrap = async ({
...
+    inputPath,
+    // if we dont batch document we can directly write to the destination folder.
+    batched ? intermediateDir : outputPath,
```
Comment (yehjxraymond) [inline src/implementations/wrap/wrap.ts:225]: Can explain this in code?

const individualDocumentsDir = batched ? intermediateDir : outputPath;

File: src/implementations/wrap/wrap.ts [hunk 6]
```diff
@@ -193,19 +219,40 @@ export const wrap = async ({
...
+    batched ? Output.Directory : outputPathType,
+    schema
+  );
 
   if (!individualDocumentHashes || individualDocumentHashes.length === 0)
     throw new Error(`No documents found in ${inputPath}`);
 
-  // Phase 2: Efficient merkling to build hashmap
-  const hashMap = merkleHashmap(individualDocumentHashes);
+  let merkleRoot: string | undefined;
```
Comment (yehjxraymond) [inline src/implementations/wrap/wrap.ts:235]: Is `| undefined` needed?

File: src/implementations/wrap/wrap.ts [hunk 7]
```diff
@@ -24,9 +24,10 @@ export enum Output {
 
 export const digestDocument = async (
   undigestedDocumentPath: string,
-  digestedDocumentDir: string,
+  digestedDocumentDir: string | undefined,
```
Comment (yehjxraymond) [inline src/implementations/wrap/wrap.ts:27]: Why it can be undefined?

File: src/implementations/wrap/wrap.ts (hunk 3 above)
Comment (Nebulis) [reply src/implementations/wrap/wrap.ts:236]: Return what?

File: src/implementations/wrap/wrap.ts (hunk 5 above)
Comment (Nebulis) [reply src/implementations/wrap/wrap.ts:225]: The current comment doesn't help ? I didn't explain the second part though. If not batched need to write to a temp folder for later to compute the melee root

File: src/implementations/wrap/wrap.ts (hunk 3 above)
Comment (Nebulis) [reply src/implementations/wrap/wrap.ts:236]: I will rephrase what do you want to return

File: src/implementations/wrap/wrap.ts (hunk 7 above)
Comment (Nebulis) [reply src/implementations/wrap/wrap.ts:27]: can be undefined if output is console

File: src/implementations/wrap/wrap.ts (hunk 4 above)
Comment (Nebulis) [reply src/implementations/wrap/wrap.ts:222]: renamed to `wrapIndividualDocuments`

File: src/implementations/wrap/wrap.ts (hunk 3 above)
Comment (Nebulis) [reply src/implementations/wrap/wrap.ts:236]: will leave like this

File: src/implementations/wrap/wrap.ts (hunk 2 above)
Comment (Nebulis) [reply src/implementations/wrap/wrap.ts:253]: removed

**Pull request merged**